# Tasca S12.01. APIs REST

# Nivell 1

## 1. Consulta l’API pública JSONPlaceholder utilitzant el mètode GET per obtenir:

Llista de publicacions (/posts)
Llista d’usuaris (/users)
Llista de tasques (/todos)

In [1]:
import requests
from requests.exceptions import HTTPError
import pandas as pd
import json

En primer lloc, com sempre, importem les llibreries de python

In [2]:
url1 = "https://jsonplaceholder.typicode.com/"
response_posts = requests.get(f"{url1}/posts")
response_posts.raise_for_status()
posts = response_posts.json()
print(posts)

[{'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}, {'userId': 1, 'id': 2, 'title': 'qui est esse', 'body': 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'}, {'userId': 1, 'id': 3, 'title': 'ea molestias quasi exercitationem repellat qui ipsa sit aut', 'body': 'et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut'}, {'userId': 1, 'id': 4, 'title': 'eum et est occaecati', 'body': 'ullam et saepe reiciendis voluptatem adipisci\nsit amet autem assumenda provident rerum culpa\nquis hic c

Probem que la conexió amb la url sigui correcte. I covertim el resultat de GET a json, per a mostrar-lo com a llista. 
Malgrat que l'exercisi no ho demana, imprimim els resultats per fer un primer cop d'ull a les dades adquirides amb JSON

In [3]:
response_users = requests.get(f"{url1}/users")
response_users.raise_for_status()
users= response_users.json()
print(users)
response_tasques = requests.get(f"{url1}/todos")
response_tasques.raise_for_status()
tasques = response_tasques.json()
print(tasques)

[{'id': 1, 'name': 'Leanne Graham', 'username': 'Bret', 'email': 'Sincere@april.biz', 'address': {'street': 'Kulas Light', 'suite': 'Apt. 556', 'city': 'Gwenborough', 'zipcode': '92998-3874', 'geo': {'lat': '-37.3159', 'lng': '81.1496'}}, 'phone': '1-770-736-8031 x56442', 'website': 'hildegard.org', 'company': {'name': 'Romaguera-Crona', 'catchPhrase': 'Multi-layered client-server neural-net', 'bs': 'harness real-time e-markets'}}, {'id': 2, 'name': 'Ervin Howell', 'username': 'Antonette', 'email': 'Shanna@melissa.tv', 'address': {'street': 'Victor Plains', 'suite': 'Suite 879', 'city': 'Wisokyburgh', 'zipcode': '90566-7771', 'geo': {'lat': '-43.9509', 'lng': '-34.4618'}}, 'phone': '010-692-6593 x09125', 'website': 'anastasia.net', 'company': {'name': 'Deckow-Crist', 'catchPhrase': 'Proactive didactic contingency', 'bs': 'synergize scalable supply-chains'}}, {'id': 3, 'name': 'Clementine Bauch', 'username': 'Samantha', 'email': 'Nathan@yesenia.net', 'address': {'street': 'Douglas Exten

Repetim el procés, aquest cop amb els usuaris i les tasques, canviant el parametre d'obtenció.

## 2. Mostra per pantalla:

La quantitat total de cada recurs.
El codi d'estat de cada petició.

In [4]:
print("La quantitat del recurs posts es de", len(posts))
print("El codi d'estat per a users es", response_posts.status_code)
print("La quantitat del recurs users es de", len(users))
print("El codi d'estat per a posts es", response_users.status_code)
print("La quantitat del recurs tasques es de", len(tasques))
print("El codi d'estat per a tasques es", response_tasques.status_code)

La quantitat del recurs posts es de 100
El codi d'estat per a users es 200
La quantitat del recurs users es de 10
El codi d'estat per a posts es 200
La quantitat del recurs tasques es de 200
El codi d'estat per a tasques es 200


Amb len() obtenim la quantitat total d'elements de cada recurs. Amb status_code accedim al codi d'estat de les peticions.

## 3. Fes una petició a una publicació inexistent per obtenir un error 404 i mostra el codi d'estat rebut.

In [5]:
response_error = requests.get(f"{url1}/patates")
try:
    response_error.raise_for_status()
except requests.exceptions.HTTPError as e:
        print(f"Error HTTP: {e}")

Error HTTP: 404 Client Error: Not Found for url: https://jsonplaceholder.typicode.com//patates


Fem de nou la petició amb una URL inventada per a obtenir l'error, amb un control Try/except per que ens retorni l'error de manera controlada i llegible.

## 4. Fes una petició POST per crear una nova publicació fictícia. Inclou el títol, el cos del missatge i un userId.

Mostra la resposta JSON.
Mostra el codi d'estat.

In [6]:
peticio_post = {'userId': "BBB", 
                'title': 'Un post de prova', 
                'body': 'Aquest és un post de prova\nPertany al exercisi 4 del nivell 1\nIT Academy - Data Analitycs\nBarcelona Activa'}

response_post = requests.post(f"{url1}/posts", json = peticio_post)
try:
    response_post.raise_for_status()
except requests.exceptions.HTTPError as e:
        print(f"Error HTTP: {e}")

post = response_post.json()
print(post)

print("El estat de la petició es", response_post.status_code)


{'userId': 'BBB', 'title': 'Un post de prova', 'body': 'Aquest és un post de prova\nPertany al exercisi 4 del nivell 1\nIT Academy - Data Analitycs\nBarcelona Activa', 'id': 101}
El estat de la petició es 201


Llencem la petició per afegir noves dades amb post, i aprofitem el codi per al control d'errors. La importació és correcta i podem veure les dades en forma de diccionari. Observem que ha assignat una id 101 (el maxim que ja existia +1). El valor de codi de estat 201 confirma que la operació s'ha fet correctament.

## 5. Fes una petició PATCH per modificar parcialment una publicació existent.

Mostra la resposta JSON.
Mostra el codi d'estat.

In [7]:
peticio_patch = {'userId': "BBB", 
                'title': 'Un post de prova modificat', 
                'body': 'Aquest és un post de prova\nPertany al exercisi 4 del nivell 1\nIT Academy - Data Analitycs 2026\nBarcelona Activa'}

response_patch = requests.patch(f"{url1}/posts/101", json = peticio_patch)
try:
    response_patch.raise_for_status()
except requests.exceptions.HTTPError as e:
        print(f"Error HTTP: {e}")

patch = response_post.json()
print(patch)

print("El estat de la petició es", response_patch.status_code)

{'userId': 'BBB', 'title': 'Un post de prova', 'body': 'Aquest és un post de prova\nPertany al exercisi 4 del nivell 1\nIT Academy - Data Analitycs\nBarcelona Activa', 'id': 101}
El estat de la petició es 200


In [8]:
peticio_patch = {'userId': "BBB", 
                'title': 'Un post de prova modificat', 
                'body': 'Aquest és un post de prova\nPertany al exercisi 4 del nivell 1\nIT Academy - Data Analitycs 2026\nBarcelona Activa'}

response_patch = requests.patch(f"{url1}/posts/1", json = peticio_patch)
try:
    response_patch.raise_for_status()
except requests.exceptions.HTTPError as e:
        print(f"Error HTTP: {e}")

patch = response_patch.json()
print(patch)

print("El estat de la petició es", response_patch.status_code)

{'userId': 'BBB', 'id': 1, 'title': 'Un post de prova modificat', 'body': 'Aquest és un post de prova\nPertany al exercisi 4 del nivell 1\nIT Academy - Data Analitycs 2026\nBarcelona Activa'}
El estat de la petició es 200


Primer, hem intentat modificar el post fictici creat en l'exercisi anterior. Però JSONplaceholder només simula la creació d'un nou recurs, i no l'hem pogut modificar, tot i no retornar error. 

A continuació, repetim la petició patch, aquest cop amb l'objectiu de modificar el post el recurs /posts/1. En aquest cas, si retorna una resposta simulada on s'han modificat les dades (com a mínim aparentment).

## 6. Fes una petició DELETE sobre una publicació.

Mostra la resposta JSON.
Mostra el codi d'estat.

In [9]:
response_delete = requests.delete(f"{url1}/posts/1")
try:
    response_delete.raise_for_status()
except requests.exceptions.HTTPError as e:
        print(f"Error HTTP: {e}")

delete = response_delete.json()
print(delete)

print("El estat de la petició es", response_delete.status_code)

{}
El estat de la petició es 200


En aquest cas fem la petició DELETE sobre el mateix conjunt de dades.

La resposta JSON és un diccionari buit {}, i el codi d’estat és 200 OK, indicant que la petició s’ha processat correctament, encara que no s’hagi eliminat cap dada real.


# Nivell 2

Interacció amb una API pública real
## 1. Explora el repositori de Public APIs i tria una API que permeti fer peticions GET.

In [10]:
swapi = "https://www.swapi.tech/api/"
response_swapi = requests.get(f"{swapi}")
try:
    response_swapi.raise_for_status()
except requests.exceptions.HTTPError as e:
        print(f"Error HTTP: {e}")
swapi_json = response_swapi.json()
print(swapi_json)



{'message': 'ok', 'result': {'films': 'https://www.swapi.tech/api/films', 'people': 'https://www.swapi.tech/api/people', 'planets': 'https://www.swapi.tech/api/planets', 'species': 'https://www.swapi.tech/api/species', 'starships': 'https://www.swapi.tech/api/starships', 'vehicles': 'https://www.swapi.tech/api/vehicles'}, 'apiVersion': '1.0', 'timestamp': '2026-05-24T10:28:38.744Z', 'support': {'contact': 'admin@swapi.tech', 'donate': 'https://www.paypal.com/donate/?business=2HGAUVTWGR5T2&no_recurring=0&item_name=Support+Swapi+and+keep+the+galaxy%27s+data+free%21+Your+donation+fuels+open-source+innovation+and+helps+us+grow.+Thank+you%21+%F0%9F%9A%80&currency_code=USD', 'partnerDiscounts': {'saberMasters': {'link': 'https://www.swapi.tech/partner-discount/sabermasters-swapi', 'details': 'Use this link to automatically get $10 off your purchase!'}, 'heartMath': {'link': 'https://www.heartmath.com/ryanc', 'details': 'Looking for some Jedi-like inner peace? Take 10% off your heart-brain co

Escollim https://www.swapi.tech/ (base de dades de l'univers Star Wars) com a objectiu de les nostres peticions. La petició GET sobre l'identificador /films ens retorna un llistat de les diferents seccións que integren la base de dades:

## 2. Llegeix la documentació de l'API:

### Revisa la secció d’endpoints disponibles. En un markdown, apunta com a mínim dos endpoints diferents.

Prendrem dos endpoint dels 6 que ens ofereix aquesta API:
    - /starships → https://www.swapi.tech/api/starships
    - /vehicles → https://www.swapi.tech/api/vehicles

### Comprova si ofereix filtres o paràmetres opcionals interessants. En un markdown, anota’ls.

In [42]:
response_swapi = requests.get(f"{swapi}/starships")
try:
    response_swapi.raise_for_status()
except requests.exceptions.HTTPError as e:
        print(f"Error HTTP: {e}")
swapi_json = response_swapi.json()
print(swapi_json)


{'message': 'ok', 'total_records': 36, 'total_pages': 4, 'previous': None, 'next': 'https://www.swapi.tech/api/starships?page=2&limit=10', 'results': [{'uid': '2', 'name': 'CR90 corvette', 'url': 'https://www.swapi.tech/api/starships/2'}, {'uid': '3', 'name': 'Star Destroyer', 'url': 'https://www.swapi.tech/api/starships/3'}, {'uid': '5', 'name': 'Sentinel-class landing craft', 'url': 'https://www.swapi.tech/api/starships/5'}, {'uid': '9', 'name': 'Death Star', 'url': 'https://www.swapi.tech/api/starships/9'}, {'uid': '11', 'name': 'Y-wing', 'url': 'https://www.swapi.tech/api/starships/11'}, {'uid': '10', 'name': 'Millennium Falcon', 'url': 'https://www.swapi.tech/api/starships/10'}, {'uid': '13', 'name': 'TIE Advanced x1', 'url': 'https://www.swapi.tech/api/starships/13'}, {'uid': '15', 'name': 'Executor', 'url': 'https://www.swapi.tech/api/starships/15'}, {'uid': '12', 'name': 'X-wing', 'url': 'https://www.swapi.tech/api/starships/12'}, {'uid': '17', 'name': 'Rebel transport', 'url':

Podem consultar els paràmetres disponibles a la documentació oficial de l'API:

https://www.swapi.tech/documentation#params

A la documentació veiem uns paràmetres generals i atributs que es poden aplicar a cada col·lecció, com per exemple "?page=" per a navegar entre els resultats, !?limit=" per limitar el nombre de resultats retornat iparàmetres especifics per a filtrar atributs concrets com "/people?name=", que ens permet buscar per nom, en el cas dels personatges:

https://www.swapi.tech/api/people?page=2

https://www.swapi.tech/api/planets?limit=5

https://www.swapi.tech/api/people?name=skywalker


### Verifica que la resposta sigui en format JSON (és el més adequat per a aquest exercici)

In [43]:
response = requests.get(f"{swapi}/starships?name=falcon")
print(response.headers["Content-Type"])

application/json; charset=utf-8


Per a verificar el format de sortida del retorn de dades es JSON, fem una petició GET a un dels endpoint, revisant els headers amb ["content_Type"]. La base de la URL "https://www.swapi.tech/api/" retorna HTLM perque es una pàgina informativa, pero els endpoint de dades si que son JSON.


## 3. Fes una petició GET senzilla:

### Mostra el codi d'estat de la resposta.

In [44]:
response_falcon = requests.get(f"{swapi}/starships?name=falcon")
try:
    response_falcon.raise_for_status()
except requests.exceptions.HTTPError as e:
        print(f"Error HTTP: {e}")

data = response_falcon.json()
print(data)

print("El estat de la petició es", response_falcon.status_code)

{'message': 'ok', 'result': [{'properties': {'created': '2026-05-23T19:51:37.622Z', 'edited': '2026-05-23T19:51:37.622Z', 'consumables': '2 months', 'name': 'Millennium Falcon', 'cargo_capacity': '100000', 'passengers': '6', 'max_atmosphering_speed': '1050', 'crew': '4', 'length': '34.37', 'model': 'YT-1300 light freighter', 'cost_in_credits': '100000', 'manufacturer': 'Corellian Engineering Corporation', 'pilots': ['https://www.swapi.tech/api/people/13', 'https://www.swapi.tech/api/people/14', 'https://www.swapi.tech/api/people/25', 'https://www.swapi.tech/api/people/31'], 'MGLT': '75', 'starship_class': 'Light freighter', 'hyperdrive_rating': '0.5', 'films': ['https://www.swapi.tech/api/films/1', 'https://www.swapi.tech/api/films/2', 'https://www.swapi.tech/api/films/3'], 'url': 'https://www.swapi.tech/api/starships/10'}, '_id': '5f63a34fee9fd7000499be23', 'description': 'A Starship', 'uid': '10', '__v': 2}], 'apiVersion': '1.0', 'timestamp': '2026-05-24T11:13:08.470Z', 'support': {'

### Imprimeix de forma clara alguns dels camps de la resposta JSON.

In [46]:
falcon_url = "https://www.swapi.tech/api/starships/10"
response_falcon = requests.get(falcon_url)
try:
    response_falcon.raise_for_status()
except requests.exceptions.HTTPError as e:
        print(f"Error HTTP: {e}")

falcon = response_falcon.json()
falcon = falcon['result']['properties']
print(falcon)

print("El estat de la petició es", response_falcon.status_code)


print("\n"+f'El fabricant de {falcon["name"]} és {falcon["manufacturer"]}')
print(f'El model és {falcon["model"]},')
print(f'Apareix en {len(falcon["films"])} pelicules de la saga')

{'created': '2026-05-23T19:51:37.622Z', 'edited': '2026-05-23T19:51:37.622Z', 'consumables': '2 months', 'name': 'Millennium Falcon', 'cargo_capacity': '100000', 'passengers': '6', 'max_atmosphering_speed': '1050', 'crew': '4', 'length': '34.37', 'model': 'YT-1300 light freighter', 'cost_in_credits': '100000', 'manufacturer': 'Corellian Engineering Corporation', 'pilots': ['https://www.swapi.tech/api/people/13', 'https://www.swapi.tech/api/people/14', 'https://www.swapi.tech/api/people/25', 'https://www.swapi.tech/api/people/31'], 'MGLT': '75', 'starship_class': 'Light freighter', 'hyperdrive_rating': '0.5', 'films': ['https://www.swapi.tech/api/films/1', 'https://www.swapi.tech/api/films/2', 'https://www.swapi.tech/api/films/3'], 'url': 'https://www.swapi.tech/api/starships/10'}
El estat de la petició es 200

El fabricant de Millennium Falcon és Corellian Engineering Corporation
El model és YT-1300 light freighter,
Apareix en 3 pelicules de la saga


Per a accedir a les caraterìstique en especific de l'item cercat, hem de feu una nova petició GET a l'URL que ens ha mostrat el resultat de la cerca de l'apartat anterior. Així, ja podrem accedir als valors de cada característica, a través de les seves claus.

## 4. Converteix les dades a un DataFrame de pandas:

Mostra les primeres files del DataFrame.

In [47]:
df_falcon = pd.DataFrame([falcon])
df_falcon

,created,edited,consumables,name,cargo_capacity,passengers,max_atmosphering_speed,crew,length,model,cost_in_credits,manufacturer,pilots,MGLT,starship_class,hyperdrive_rating,films,url
0,2026-05-23T19:51:37.622Z,2026-05-23T19:51:37.622Z,2 months,Millennium Falcon,100000,6,1050,4,34.37,YT-1300 light freighter,100000,Corellian Engineering Corporation,"[https://www.swapi.tech/api/people/13, https:/...",75,Light freighter,0.5,"[https://www.swapi.tech/api/films/1, https://w...",https://www.swapi.tech/api/starships/10


Per a convertir el resultat de la cerca en un DataFrame, hem de passar el valor entre "[]", per convertirlo a un diccionari que agafa la forma de DataFrame d'una sola columna. Sense això, la diferent longituts dels arrays dels resultats, no permeten la conversió.

# Nivell 3

## 1. Utilitza l’API d’Open Data BCN per:

Cercar un dataset del teu interès mitjançant package_search i package_show.

In [16]:
bcn_od_url = "https://opendata-ajuntament.barcelona.cat/data/api/action/"
response_bcn = requests.get(f"{bcn_od_url}//package_list")
try:
    response_bcn.raise_for_status()
except requests.exceptions.HTTPError as e:
        print(f"Error HTTP: {e}")

bcn_data_list = response_bcn.json()

bcn_data_list["result"]

['20170706-districtes-barris',
 'accessibilitat-via-publica',
 'accidents_causa_conductor_gu_bcn',
 'accidents-causes-gu-bcn',
 'accidents-gu-bcn',
 'accidents-persones-gu-bcn',
 'accidents-tipus-gu-bcn',
 'accidents-vehicles-gu-bcn',
 'accions-formatives-adrecades-al-personal',
 'acollida-animals',
 'administracio-autonomica',
 'administracio-estatal',
 'administracio-municipal',
 'afectacions-turistiques',
 'afectacions-turistiques-a-espais-verds',
 'aforaments-descriptiu',
 'aforaments-detall',
 'aforament-tipus-equipament',
 'agenda-activitats-esportives',
 'agenda-cultural',
 'agenda-diaria',
 'agenda-publica-alcaldesa',
 'agendes-publiques-comissionats',
 'agendes-publiques-regidors',
 'alcaldes-alcaldesses-bcn',
 'allotjaments-altres',
 'allotjaments-hotels',
 'allotjaments-pensions',
 'ampliacio-voreres-bici-bcn',
 'aparcaments-bcn',
 'aparcaments-servei-bicis',
 'aparcaments-sota-superficie',
 'arbrat-parcs',
 'arbrat-viari',
 'arbrat-zona',
 'arbres-interes-local',
 'arees-es

En primer terme, accedim al llistat de datasets actual a través del metode "/package_list", per a escollir un dataset.

## 2. Dels resultats obtinguts, selecciona’n un que tingui recursos disponibles en CSV o JSON.

In [25]:
bcn_od_url = "https://opendata-ajuntament.barcelona.cat/data/api/action/"
response_agenda = requests.get(f"{bcn_od_url}/package_show?id=agenda-cultural")
try:
    response_agenda.raise_for_status()
except requests.exceptions.HTTPError as e:
        print(f"Error HTTP: {e}")

agenda = response_agenda.json()

agenda


{'help': 'https://opendata-ajuntament.barcelona.cat/data/api/3/action/help_show?name=package_show',
 'success': True,
 'result': {'notes_translated': {'ca': 'Activitats culturals de la ciutat de Barcelona classificada per temàtiques.  ',
   'en': 'Cultural activities of the city of Barcelona classified by subject matters. ',
   'es': 'Actividades culturales de la ciudad de Barcelona clasificadas por temáticas.'},
  'geolocation': 'Yes_with_map_view',
  'code': 'AGENDA_CULTURAL',
  'url_tornada': {'ca': 'http://guia.barcelona.cat/agenda',
   'en': 'http://guia.barcelona.cat/en/agenda/',
   'es': 'http://guia.barcelona.cat/es/agenda/'},
  'dataset_fields_description': '__Important:__ A partir del __24/03/2021__  i en relació a canvis en el sistema origen de les dades, es deixa de publicar el recurs en format .xml quedant substituït per el de format .json.',
  'fuente': 'Ajuntament de Barcelona',
  'private': False,
  'num_tags': 2,
  'api': 'Yes',
  'frequency': 'DIARIA',
  'id': '276715

Ente el llistat, trobem agenda-cultural, que disposa de recursos en format .csv. Busquem el resource id que mostra el json del recurs:

"id': '3abb2414-1ee0-446e-9c25-380e938adb73"

## 3. Amb el resource_id del recurs seleccionat, realitza una consulta mitjançant datastore_search per:

Recuperar almenys 100 registres.

In [33]:
resource_id = "3abb2414-1ee0-446e-9c25-380e938adb73"

url_search = f"{bcn_od_url}/datastore_search?resource_id={resource_id}&limit=200"

response_search = requests.get(url_search)
response_search.raise_for_status()

data_search = response_search.json()
data_search["result"]["records"]

[{'addresses_roadtype_name': None,
  'addresses_end_street_number': '',
  'geo_epgs_4326_lon': None,
  'values_attribute_name': '',
  'estimated_dates': '',
  'addresses_road_name': '',
  'values_category': '',
  'addresses_zip_code': '',
  'secondary_filters_id': '',
  'values_value': '',
  'addresses_town': 'Barcelona',
  'secondary_filters_name': '',
  'timetable': None,
  'secondary_filters_tree': '',
  'start_date': '2026-04-23T03:00:00+02:00',
  'addresses_district_name': '',
  'end_date': '2026-10-10T03:00:00',
  'geo_epgs_25831_x': None,
  'addresses_start_street_number': '',
  'register_id': '\ufeff99400654628',
  'institution_id': '',
  'addresses_main_address': 'True',
  'addresses_district_id': '',
  'addresses_roadtype_id': None,
  'addresses_type': '',
  'addresses_neighborhood_id': '',
  '_id': 1,
  'name': '"Cicle d\'espectacles i concerts ""Terrats en Cultura"""',
  'addresses_road_id': '',
  'created': '2020-07-06T10:02:55+02:00',
  'geo_epgs_25831_y': None,
  'instit

Configurem la cerca amb l'ID del dataset i hi posem un limit de 200, en observar que el resultat conté molts camps buits, per tal de poder tenir una major visió del tipus de dades que trobarem en aquest dataset. 

Desprès en filtrem els resultats per a confeccionar el DataFrame i posteriorment, exportar com a CSV.

## 4. Converteix els resultats a un DataFrame.

In [31]:
df = pd.DataFrame(data_search["result"]["records"])
df


,addresses_roadtype_name,addresses_end_street_number,geo_epgs_4326_lon,values_attribute_name,estimated_dates,addresses_road_name,values_category,addresses_zip_code,secondary_filters_id,values_value,...,institution_name,modified,secondary_filters_asia_id,secondary_filters_fullpath,geo_epgs_4326_lat,values_description,values_id,addresses_neighborhood_name,values_outstanding,values_attribute_id
0,None,,None,,,,,,,,...,,2026-05-15T08:15:50.159055+02:00,,,None,,,,,
1,None,,2.1564074948215346,,,Gran Via de les Corts Catalanes,,08015,,,...,,2026-03-02T15:21:27.162991+01:00,,,41.3808491471351,,,la Nova Esquerra de l'Eixample,,
2,None,,2.165900883899018,,,Carrer de València,,08009,,,...,,2026-03-10T11:11:24.540789+01:00,,,41.394518975814755,,,la Dreta de l'Eixample,,
3,None,,2.165900883899018,,,Carrer de València,,08009,,,...,,2026-03-10T13:00:31.400168+01:00,,,41.394518975814755,,,la Dreta de l'Eixample,,
4,None,,2.1251516661411447,,,C Major de Sarrià,,08017,,,...,,2026-04-10T13:44:48.834590+02:00,,,41.39599253027976,,,Sarrià,,
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,None,,2.182465058010549,,,Plaça de Fort Pienc,,08013,,,...,,2026-03-05T09:46:16.925616+01:00,,,41.39558658720727,,,el Fort Pienc,,
96,None,,2.184931573020558,,,Carrer de Las Navas de Tolosa,,08027,,,...,,2026-03-12T14:53:22.614247+01:00,,,41.41693998158291,,,Navas,,
97,None,,2.182465058010549,,,Plaça de Fort Pienc,,08013,,,...,,2026-03-05T10:49:35.710589+01:00,,,41.39558658720727,,,el Fort Pienc,,
98,None,2,4582085.127,20055,41.387339989243294,301506,612577,01,False,Telèfons,...,,2024-07-01T12:03:10.425706+02:00,,,431037.525,932 957 207,True,04,Venda d'entrades,


## 5. Desa el DataFrame en un fitxer .csv.

In [34]:
df.to_csv("agenda_cultural.csv")